### DISCUSSION TITLE
Delete `refs/pull/NUMBER/head` Git ref when PR is closed/merged

### DISCUSSION THREAD

**Discussion by jmarshall** [community/community#11922] [Pull Requests]
As somewhat obliquely documented in various _docs.github.com_ pages, GitHub maintains two Git refs for each pull request:

* `refs/pull/NUMBER/head` refers to the head commit of pull request \#NUMBER
* `refs/pull/NUMBER/merge` refers to a trial merge of pull request \#NUMBER into its base branch

As discussed in [this StackOverflow answer](https://stackoverflow.com/questions/28103251/check-out-a-pull-request-on-git-to-review-test-a-pr/28106660#28106660) and in GitHub's [Modifying PRs locally documentation](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally#modifying-an-inactive-pull-request-locally), these refs can be used to easily fetch a PR's commits into a local repository:

```
$ git fetch origin pull/NUMBER/head:BRANCHNAME
```

One could also set up a repository's _.git/config_ so that these refs are fetched automatically (for _all_ pull requests) by `git fetch`:

```
[remote "origin"]
  fetch = +refs/heads/*:refs/remotes/origin/*
  url = git@github.com:<USERNAME>/<REPO_NAME>.git
  fetch = +refs/pull/*/head:refs/remotes/origin/pr/*    # ADD THIS LINE TO THE USUAL SETTINGS
```

This is discussed in various blog postings (e.g. [this tutorial](https://github.com/Spredzy/productive-git/blob/master/src/collaboration/github-pr-locally.md), [this gist](https://gist.github.com/piscisaureus/3342247), and [this posting](https://www.jvt.me/posts/2019/01/19/git-ref-github-pull-requests/)), [this StackOverflow question](https://stackoverflow.com/questions/17644164/refs-pull-head-origin-pr-appearing-on-a-git-pull), and the [same GitHub documentation page until 2014](https://web.archive.org/web/20130816145110/https://help.github.com/articles/checking-out-pull-requests-locally).

This is a very useful way of staying up-to-date with pull requests on a repository that you follow closely. However what is less useful is that these refs accumulate, staying around even years after a PR has been merged or otherwise closed. (You could delete them locally, but they will just reappear the next time you `git fetch` from that remote.)

What would make this even more useful would be if the `refs/pull/NUMBER/head` ref was deleted from the GitHub repository when the PR was merged or closed. Then when the user runs `git fetch --prune` their local refs to PRs will (a) spring into existence when a new PR is opened; (b) be updated if an active PR is pushed to; (c) disappear when the PR is no longer active.

This would mean that the local repository would only have these PR-related refs for currently active PRs; listings would not be cluttered by long-since inactive PRs.

It may not be desired to change the semantics of `refs/pull/NUMBER/head` in this way. If so, an attractive alternative would be to introduce a third GitHub-maintained ref for each pull request:

* `refs/pull/NUMBER/active` would be the same as `refs/pull/NUMBER/head` when pull request \#NUMBER is open, and would be deleted when the PR is merged or otherwise closed

This could be named e.g. `…/active` or `…/open` or `…/open_head` or similar.

Then users could add a `fetch = +refs/pull/*/active:…` (or whatever name is chosen) instead of the `fetch` line show above, and `git fetch -p` would automatically maintain local refs providing easy access to the set of currently active PRs.

**Comment by xmo-odoo** [comment]
> What would make this even more useful would be if the refs/pull/NUMBER/head ref was deleted from the GitHub repository when the PR was merged or closed. [...] It may not be desired to change the semantics of `refs/pull/NUMBER/head` in this way.

I think it is not as that ref is valuable both before and after merging the PR e.g. if tooling wants to perform a backport or forwardport of the merged PR.

It also doesn't match the semantics of the name: the PR still has a head even if it's been merged/closed.

**Reply by jmarshall** [reply]
Indeed, those may be considerations. That is why I included the alternative of leaving `refs/pull/NUMBER/head` as is and adding a third ref that would provide the proposed desired functionality on fetching.

Do you agree that having a set of refs that corresponds to the set of active (i.e., open) PRs would be useful?

**Reply by xmo-odoo** [reply]
I can't say it would be useful to me, but I can certainly imagine that it would be useful to other people.

**Comment by ljharb** [comment]
I would NOT want these refs to be automatically deleted, but I *absolutely* would love to be able TO delete them.

This is also an issue when a spam PR is deleted by github; the PR ref remains forever.

**Reply by jmarshall** [reply]
Once again, that is why the proposal includes an alternative of adding a new `…/active` ref and only deleting that one.

Perhaps you could say more about why you think the existing refs should not be automatically deleted.

**Reply by ljharb** [reply]
Because one of the primary values for me is the programmatically detectable PR ref for merged PRs.

I’d went the ability to delete things that weren’t reachable from the default branch, but I’d want to *prevent* deletion of anything that is reachable from the default branch.
